## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [711]:
import pandas as pd

custumer_df = pd.read_csv('starter_code/Resources/client_dataset.csv')

custumer_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [712]:
# View the column names in the data
custumer_df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [713]:
# Use the describe function to gather some basic statistics
custumer_df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [714]:
# Use this space to do any additional research
# and familiarize yourself with the data.
outlier_df = custumer_df.loc[(custumer_df['order_year'] < 2000) |
                                (custumer_df['unit_cost'] > 500)]
outlier_df.head()


first     last                         job                 phone  \
0       Donald  Harding                Immunologist    793-904-7725x39308   
95        Erik    Gross                  IT trainer     313-970-3023x5865   
110   Victoria   Flores    Educational psychologist      001-418-361-8621   
118     Ashley      Kim    Local government officer  001-895-417-4003x017   
225  Alexander    Blake  Museum/gallery conservator    315-727-3712x50544   

                                   email  client_id  order_id  order_date  \
0       harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
95                  erikgross@nelson.com      32098   3998530  2023-05-20   
110  victoriainternational@blanchard.com      65049   1720923  2023-05-04   
118              kimashley@stevenson.net      53348   5741617  2023-04-08   
225             alexander.blake@lang.com      94174   3294118  2023-03-24   

     order_week  order_year         item_id     category       subcategory  \
0            17        2023  EUD29711-63-6U        decor          wall art   
95           20        2023  EUD29711-63-6U        decor          wall art   
110          18        2023  ZAM61851-20-6X  electronics           telecom   
118          14        2023  EUD29711-63-6U        decor          wall art   
225          12        2023  DBC69293-78-9M  consumables  kitchen supplies   

     unit_price  unit_cost  unit_weight  qty  line_number  
0       1096.80     762.71         7.50  105            1  
95      1096.80     762.71         7.50  107            3  
110      784.95     684.06         3.04   22            0  
118     1096.80     762.71         7.50   73            6  
225     1132.23     628.35         3.02    9            3

In [715]:
len(outlier_df)

1196

In [716]:
outlier1_df = custumer_df.loc[(custumer_df['order_year'] < 2000) |
                                (custumer_df['unit_cost'] > 800)]
outlier1_df.head()

first     last                                 job  \
675      Julie   Acosta                            Musician   
1136    Walter     Tran  Public relations account executive   
1901   Richard   Murphy                 Librarian, academic   
2178    Ashley  Krueger          Research scientist (maths)   
3135  Jonathan  Kennedy                    Textile designer   

                     phone                        email  client_id  order_id  \
675    +1-664-222-1815x094           acostaj@taylor.com      50605   3263138   
1136     509.895.5322x0674      tranw@cain-williams.com      16624   7323688   
1901  001-506-286-9298x756  murphy.richard@elliott.info      76418   3065418   
2178          747.503.1276      ashleykrueger@terry.com      43576   1321206   
3135      832.240.4049x974         kennedy_j@murray.com      52670   8317238   

      order_date  order_week  order_year         item_id     category  \
675   2023-01-17           3        2023  BBE61778-67-4A  consumables   
1136  2023-02-14           7        2023  BBE61778-67-4A  consumables   
1901  2023-05-30          22        2023  BBE61778-67-4A  consumables   
2178  2023-03-21          12        2023  BBE61778-67-4A  consumables   
3135  2023-05-15          20        2023  BBE61778-67-4A  consumables   

            subcategory  unit_price  unit_cost  unit_weight  qty  line_number  
675   bathroom supplies     1396.23     846.27         9.53  115            5  
1136  bathroom supplies     1396.23     846.27         9.53   94            1  
1901  bathroom supplies     1396.23     846.27         9.53   35            6  
2178  bathroom supplies     1396.23     846.27         9.53   96            7  
3135  bathroom supplies     1396.23     846.27         9.53   27            2

In [717]:
custumer_df.count()

first          54639
last           54639
job            54639
phone          54639
email          54639
client_id      54639
order_id       54639
order_date     54639
order_week     54639
order_year     54639
item_id        54639
category       54639
subcategory    54639
unit_price     54639
unit_cost      54639
unit_weight    54639
qty            54639
line_number    54639
dtype: int64

In [718]:
len(outlier_df)

1196

In [719]:
# What three item categories had the most entries?
category_counts = custumer_df['category'].value_counts()
top_three_categories = category_counts.head(3)
print("Top Three Categories with Most Entries:")
print(top_three_categories)

Top Three Categories with Most Entries:
category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [720]:
# For the category with the most entries,
# which subcategory had the most entries?

#top_three_categories = df.groupby('category')['subcategory'].value_counts()

most_sub_category = custumer_df.groupby('category')['subcategory'].value_counts()
most_sub_category


category     subcategory       
consumables  bathroom supplies     6424
             paper products        5874
             kitchen supplies      5746
             misc                  2891
             pens                  2603
decor        plants                3466
             wall art              3093
             misc                  1395
electronics  telecom                978
             audio visual           732
             cabling                655
             computers              467
furniture    tables                4540
             stands                4201
             seating               2664
             desks                  510
software     accounting            3756
             project management    3522
             logistics             1122
Name: count, dtype: int64

In [721]:
# Which five clients had the most entries in the data?
main_clients = custumer_df['client_id'].value_counts().sort_values(ascending=False)
main_clients.head()

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [722]:
# Store the client ids of those top 5 clients in a list.
#five_clients_id = five_clients_id.nlargest(5).index.tolist()
#print("Your Top 5 Clients are:", five_clients_id)
top_five_clients = list(custumer_df['client_id'].value_counts().head(5).index)
top_five_clients

[33615, 66037, 46820, 38378, 24741]

In [723]:
# How many total units (the qty column) did the
# client with the most entries order order?
client_most_entries_df = custumer_df[custumer_df['client_id'] == 33615]
total_units = client_most_entries_df['qty'].sum()
total_units


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [724]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
subtotal_df = custumer_df
subtotal_df['subtotal'] = subtotal_df['unit_price'] * subtotal_df['qty']
subtotal_df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number   subtotal  
0                1  115164.00  
1                0     523.95  
2                6     527.28  
3                3    1056.18  
4                1    3902.00  
...            ...        ...  
54634            8    2743.29  
54635            0    9709.73  
54636            6   31188.50  
54637            0     165.76  
54638            5    3103.31  

[54639 rows x 19 columns]

In [748]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
def shipping_rate(weight):
    if weight > 50:
        return weight * 7
    return weight * 10

subtotal_df['total_weight'] = subtotal_df['qty'] * subtotal_df['unit_weight']
subtotal_df['shipping price'] = subtotal_df['total_weight'].apply(shipping_rate)
subtotal_df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping price']].head(5)


unit_price  unit_weight  qty  total_weight  shipping price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20

In [749]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
subtotal_df['total price'] = subtotal_df['subtotal'] * 1.0925 + subtotal_df['shipping price']
subtotal_df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
0              17        2023  ...    762.71        7.50   105            1   
1              20        2023  ...     15.09        1.49    21            0   
2               4        2023  ...      7.86        1.68    39            6   
3              17        2023  ...     24.85        1.23    29            3   
4              21        2023  ...    108.17       46.43    20            1   
...           ...         ...  ...       ...         ...   ...          ...   
54634          17        2023  ...     51.60        2.25    33            8   
54635           6        2023  ...    175.46       11.70    47            0   
54636           9        2023  ...     57.31        4.16   475            6   
54637           3        2023  ...      1.22       18.04   112            0   
54638           1        2023  ...      1.61        2.07  1031            5   

        subtotal  shipping price    total price      cost        profit  \
0      115164.00         5512.50  131329.170000  85597.05  45732.120000   
1         523.95          312.90     885.315375    629.79    255.525375   
2         527.28          458.64    1034.693400    765.18    269.513400   
3        1056.18          356.70    1510.576650   1077.35    433.226650   
4        3902.00         6500.20   10763.135000   8663.60   2099.535000   
...          ...             ...            ...       ...           ...   
54634    2743.29          519.75    3516.794325   2222.55   1294.244325   
54635    9709.73         3849.30   14457.180025  12095.92   2361.260025   
54636   31188.50        13832.00   47905.436250  41054.25   6851.186250   
54637     165.76        14143.36   14324.452800  14280.00     44.452800   
54638    3103.31        14939.19   18329.556175  16599.10   1730.456175   

       total_weight  
0            787.50  
1             31.29  
2             65.52  
3             35.67  
4            928.60  
...             ...  
54634         74.25  
54635        549.90  
54636       1976.00  
54637       2020.48  
54638       2134.17  

[54639 rows x 24 columns]

In [750]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
subtotal_df['cost'] = subtotal_df['unit_cost'] * subtotal_df['qty'] + subtotal_df['shipping price']
subtotal_df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
0              17        2023  ...    762.71        7.50   105            1   
1              20        2023  ...     15.09        1.49    21            0   
2               4        2023  ...      7.86        1.68    39            6   
3              17        2023  ...     24.85        1.23    29            3   
4              21        2023  ...    108.17       46.43    20            1   
...           ...         ...  ...       ...         ...   ...          ...   
54634          17        2023  ...     51.60        2.25    33            8   
54635           6        2023  ...    175.46       11.70    47            0   
54636           9        2023  ...     57.31        4.16   475            6   
54637           3        2023  ...      1.22       18.04   112            0   
54638           1        2023  ...      1.61        2.07  1031            5   

        subtotal  shipping price    total price      cost        profit  \
0      115164.00         5512.50  131329.170000  85597.05  45732.120000   
1         523.95          312.90     885.315375    629.79    255.525375   
2         527.28          458.64    1034.693400    765.18    269.513400   
3        1056.18          356.70    1510.576650   1077.35    433.226650   
4        3902.00         6500.20   10763.135000   8663.60   2099.535000   
...          ...             ...            ...       ...           ...   
54634    2743.29          519.75    3516.794325   2222.55   1294.244325   
54635    9709.73         3849.30   14457.180025  12095.92   2361.260025   
54636   31188.50        13832.00   47905.436250  41054.25   6851.186250   
54637     165.76        14143.36   14324.452800  14280.00     44.452800   
54638    3103.31        14939.19   18329.556175  16599.10   1730.456175   

       total_weight  
0            787.50  
1             31.29  
2             65.52  
3             35.67  
4            928.60  
...             ...  
54634         74.25  
54635        549.90  
54636       1976.00  
54637       2020.48  
54638       2134.17  

[54639 rows x 24 columns]

In [751]:
# Create a column for the profit of
# each line using line cost and line price
subtotal_df['profit'] = subtotal_df['total price'] - subtotal_df['cost']
subtotal_df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
0              17        2023  ...    762.71        7.50   105            1   
1              20        2023  ...     15.09        1.49    21            0   
2               4        2023  ...      7.86        1.68    39            6   
3              17        2023  ...     24.85        1.23    29            3   
4              21        2023  ...    108.17       46.43    20            1   
...           ...         ...  ...       ...         ...   ...          ...   
54634          17        2023  ...     51.60        2.25    33            8   
54635           6        2023  ...    175.46       11.70    47            0   
54636           9        2023  ...     57.31        4.16   475            6   
54637           3        2023  ...      1.22       18.04   112            0   
54638           1        2023  ...      1.61        2.07  1031            5   

        subtotal  shipping price    total price      cost        profit  \
0      115164.00         5512.50  131329.170000  85597.05  45732.120000   
1         523.95          312.90     885.315375    629.79    255.525375   
2         527.28          458.64    1034.693400    765.18    269.513400   
3        1056.18          356.70    1510.576650   1077.35    433.226650   
4        3902.00         6500.20   10763.135000   8663.60   2099.535000   
...          ...             ...            ...       ...           ...   
54634    2743.29          519.75    3516.794325   2222.55   1294.244325   
54635    9709.73         3849.30   14457.180025  12095.92   2361.260025   
54636   31188.50        13832.00   47905.436250  41054.25   6851.186250   
54637     165.76        14143.36   14324.452800  14280.00     44.452800   
54638    3103.31        14939.19   18329.556175  16599.10   1730.456175   

       total_weight  
0            787.50  
1             31.29  
2             65.52  
3             35.67  
4            928.60  
...             ...  
54634         74.25  
54635        549.90  
54636       1976.00  
54637       2020.48  
54638       2134.17  

[54639 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [752]:
# Check your work using the totals above
order1_verify_df = subtotal_df[subtotal_df['order_id'] == 2742071]
order1_total = order1_verify_df['total price'].sum()

order2_verify_df = subtotal_df[subtotal_df['order_id'] == 2173913]
order2_total = order2_verify_df['total price'].sum()

order3_verify_df = subtotal_df[subtotal_df['order_id'] == 6128929]
Order3_total = order3_verify_df['total price'].sum()

print(f"Order ID 2742071: $${order1_total}")
print(f"Order ID 2173913: $${order2_total}")
print(f"Order ID 6128929: $${Order3_total}")

Order ID 2742071: $$151695.028025
Order ID 2173913: $$158762.83029999997
Order ID 6128929: $$908947.334975


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [753]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
#top_quantity = subtotal_df.groupby('qty')['client_id'].value_counts()
#top_quantity.tail(5)
top_quantity_df = custumer_df[custumer_df['client_id'].apply(lambda x: x in top_five_clients)]

top_quantity_df.head(5)

first     last                                           job  \
5    Jessica    Reyes  Production designer, theatre/television/film   
12   Jessica    Reyes  Production designer, theatre/television/film   
137    Bryan    Myers                                Chief of Staff   
214    Bryan    Myers                                Chief of Staff   
216   Angela  Everett                                     Osteopath   

                phone                     email  client_id  order_id  \
5          5016915923  jessica.reyes@grimes.net      33615   1327730   
12         5016915923  jessica.reyes@grimes.net      33615   6059361   
137  444.233.9146x233     bmyers3696@wilson.com      46820   7109502   
214  444.233.9146x233     bmyers3696@wilson.com      46820   8967498   
216  729-620-6192x875    aeverett@rodriguez.com      66037   1203693   

     order_date  order_week  order_year  ... unit_cost unit_weight  qty  \
5    2023-05-02          18        2023  ...    154.44        2.79   78   
12   2023-05-16          20        2023  ...    198.38       27.51  568   
137  2023-05-12          19        2023  ...    111.47        5.65  118   
214  2023-02-06           6        2023  ...      6.26        7.05  133   
216  2023-04-09          14        2023  ...    166.09        0.26  157   

     line_number   subtotal  shipping price    total price       cost  \
5              0   13481.52         1523.34   16251.900600   13569.66   
12             2  130492.32       109379.76  251942.619600  222059.60   
137            9   16063.34         4666.90   22216.098950   17820.36   
214            3    1061.34         6563.55    7723.063950    7396.13   
216            2   47393.59          408.20   52185.697075   26484.33   

           profit  total_weight  
5     2682.240600        217.62  
12   29883.019600      15625.68  
137   4395.738950        666.70  
214    326.933950        937.65  
216  25701.367075         40.82  

[5 rows x 24 columns]

In [754]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
new_summary_df = top_quantity_df[['client_id', 'qty', 'shipping price', 'total price', 'profit']]
new_summary_df = new_summary_df.sort_values('profit', ascending=False).head(5).reset_index(drop=True)
new_summary_df


client_id     qty  shipping price   total price        profit
0      24741  122166      1590601.32  6.315863e+07  3.127819e+07
1      24741   65262      1749674.22  1.089944e+07  2.851984e+06
2      66037    3678       125897.94  2.369147e+06  6.682557e+05
3      46820    6099        22627.29  2.361595e+06  5.248206e+05
4      33615   17750       438602.50  1.763261e+06  5.168565e+05

In [755]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
new_summary_name_df = top_quantity_df.rename(columns={'client_id':'Purchased by:', 'qty': 'Amount Purchased', 'shipping price':'Shipping Fees','total price':'Final Cost', 'profit': 'Total Profit'})
new_summary_name_df


first     last                                           job  \
5        Jessica    Reyes  Production designer, theatre/television/film   
12       Jessica    Reyes  Production designer, theatre/television/film   
137        Bryan    Myers                                Chief of Staff   
214        Bryan    Myers                                Chief of Staff   
216       Angela  Everett                                     Osteopath   
...          ...      ...                                           ...   
54535     Angela  Everett                                     Osteopath   
54536     Kendra  Garrett                        Pension scheme manager   
54579  Alexandra    Young                                     Dietitian   
54589  Alexandra    Young                                     Dietitian   
54613     Kendra  Garrett                        Pension scheme manager   

                  phone                            email  Purchased by:  \
5            5016915923         jessica.reyes@grimes.net          33615   
12           5016915923         jessica.reyes@grimes.net          33615   
137    444.233.9146x233            bmyers3696@wilson.com          46820   
214    444.233.9146x233            bmyers3696@wilson.com          46820   
216    729-620-6192x875           aeverett@rodriguez.com          66037   
...                 ...                              ...            ...   
54535  729-620-6192x875           aeverett@rodriguez.com          66037   
54536      808.257.2029  garrettkendra8479@jefferson.com          24741   
54579  265-666-3155x118           ayoung9158@stevens.com          38378   
54589  265-666-3155x118           ayoung9158@stevens.com          38378   
54613      808.257.2029  garrettkendra8479@jefferson.com          24741   

       order_id  order_date  order_week  order_year  ... unit_cost  \
5       1327730  2023-05-02          18        2023  ...    154.44   
12      6059361  2023-05-16          20        2023  ...    198.38   
137     7109502  2023-05-12          19        2023  ...    111.47   
214     8967498  2023-02-06           6        2023  ...      6.26   
216     1203693  2023-04-09          14        2023  ...    166.09   
...         ...         ...         ...         ...  ...       ...   
54535   7550600  2023-02-24           8        2023  ...     10.66   
54536   1886506  2023-05-26          21        2023  ...     40.05   
54579   8661147  2023-03-09          10        2023  ...     20.29   
54589   4508675  2023-03-25          12        2023  ...     38.26   
54613   9114777  2023-02-13           7        2023  ...     59.32   

      unit_weight Amount Purchased  line_number   subtotal  Shipping Fees  \
5            2.79               78            0   13481.52        1523.34   
12          27.51              568            2  130492.32      109379.76   
137          5.65              118            9   16063.34        4666.90   
214          7.05              133            3    1061.34        6563.55   
216          0.26              157            2   47393.59         408.20   
...           ...              ...          ...        ...            ...   
54535        7.52              108            2    1810.08        5685.12   
54536        4.70              721            3   46014.22       23720.90   
54579       13.07              118            3    3687.50       10795.82   
54589        0.70              685            0   35784.40        3356.50   
54613       13.13               91            2    8705.97        8363.81   

          Final Cost       cost  Total Profit  total_weight  
5       16251.900600   13569.66   2682.240600        217.62  
12     251942.619600  222059.60  29883.019600      15625.68  
137     22216.098950   17820.36   4395.738950        666.70  
214      7723.063950    7396.13    326.933950        937.65  
216     52185.697075   26484.33  25701.367075         40.82  
...              ...        ...           ...           ...  
54535    

In [756]:
# Sort the updated data by "Total Profit" form highest to lowest
new_summary_name_df.sort_values('Total Profit', ascending=False)
new_summary_name_df

first     last                                           job  \
5        Jessica    Reyes  Production designer, theatre/television/film   
12       Jessica    Reyes  Production designer, theatre/television/film   
137        Bryan    Myers                                Chief of Staff   
214        Bryan    Myers                                Chief of Staff   
216       Angela  Everett                                     Osteopath   
...          ...      ...                                           ...   
54535     Angela  Everett                                     Osteopath   
54536     Kendra  Garrett                        Pension scheme manager   
54579  Alexandra    Young                                     Dietitian   
54589  Alexandra    Young                                     Dietitian   
54613     Kendra  Garrett                        Pension scheme manager   

                  phone                            email  Purchased by:  \
5            5016915923         jessica.reyes@grimes.net          33615   
12           5016915923         jessica.reyes@grimes.net          33615   
137    444.233.9146x233            bmyers3696@wilson.com          46820   
214    444.233.9146x233            bmyers3696@wilson.com          46820   
216    729-620-6192x875           aeverett@rodriguez.com          66037   
...                 ...                              ...            ...   
54535  729-620-6192x875           aeverett@rodriguez.com          66037   
54536      808.257.2029  garrettkendra8479@jefferson.com          24741   
54579  265-666-3155x118           ayoung9158@stevens.com          38378   
54589  265-666-3155x118           ayoung9158@stevens.com          38378   
54613      808.257.2029  garrettkendra8479@jefferson.com          24741   

       order_id  order_date  order_week  order_year  ... unit_cost  \
5       1327730  2023-05-02          18        2023  ...    154.44   
12      6059361  2023-05-16          20        2023  ...    198.38   
137     7109502  2023-05-12          19        2023  ...    111.47   
214     8967498  2023-02-06           6        2023  ...      6.26   
216     1203693  2023-04-09          14        2023  ...    166.09   
...         ...         ...         ...         ...  ...       ...   
54535   7550600  2023-02-24           8        2023  ...     10.66   
54536   1886506  2023-05-26          21        2023  ...     40.05   
54579   8661147  2023-03-09          10        2023  ...     20.29   
54589   4508675  2023-03-25          12        2023  ...     38.26   
54613   9114777  2023-02-13           7        2023  ...     59.32   

      unit_weight Amount Purchased  line_number   subtotal  Shipping Fees  \
5            2.79               78            0   13481.52        1523.34   
12          27.51              568            2  130492.32      109379.76   
137          5.65              118            9   16063.34        4666.90   
214          7.05              133            3    1061.34        6563.55   
216          0.26              157            2   47393.59         408.20   
...           ...              ...          ...        ...            ...   
54535        7.52              108            2    1810.08        5685.12   
54536        4.70              721            3   46014.22       23720.90   
54579       13.07              118            3    3687.50       10795.82   
54589        0.70              685            0   35784.40        3356.50   
54613       13.13               91            2    8705.97        8363.81   

          Final Cost       cost  Total Profit  total_weight  
5       16251.900600   13569.66   2682.240600        217.62  
12     251942.619600  222059.60  29883.019600      15625.68  
137     22216.098950   17820.36   4395.738950        666.70  
214      7723.063950    7396.13    326.933950        937.65  
216     52185.697075   26484.33  25701.367075         40.82  
...              ...        ...           ...           ...  
54535    